### ISSUE TITLE
[ansi_stripper] stripper doesn't strip color?

### ISSUE THREAD

**Issue by KannarFr** [timberio/vector#2202]
Hi,

ansi_stripper transform doesn't remove ansi color from field?

For the following test:
```
[transforms.ansi_stripper_simple_yo]
  inputs = []
  type = "ansi_stripper"
  field = "message"
[[tests]]
  name = "ansi_stripper_simple_yo"
  [tests.input]
    insert_at = "ansi_stripper_simple_yo"
    type = "raw"
    value = "\t\r\\033[31mHelloWorld"
  [[tests.outputs]]
    extract_from = "ansi_stripper_simple_yo"
    [[tests.outputs.conditions]]
      "message.equals" = "HelloWorld"
```

It outputs:
```
check transform 'ansi_stripper_simple_yo' failed conditions:
  condition[0]: predicates failed: [ message.equals: "HelloWorld" ]
payloads (events encoded as JSON):
  input: {"message":"\t\r\\033[31mHelloWorld","timestamp":"2020-04-02T03:52:42.774856684Z"}
  output: {"message":"\\033[31mHelloWorld","timestamp":"2020-04-02T03:52:42.774856684Z"}
```

Maybe it's a wanted behavior and we should write ansi_color_stripper? WDYT?

**Comment by binarylogic** [comment]
Hi @KannarFr thanks for reporting. Yes, this should probably stip color as well.  We'll get this fixed.

**Comment by binarylogic** [comment]
@JeanMertz given that you built this transform could you take a look? It should be relatively simple, but if this test case is correct this transform is not really usable. Thanks!

**Comment by KannarFr** [comment]
@JeanMertz I think there is an issue on the parsing:

```
value = "\033[31;1;4mHello\033[0m"
```

throws: 

```
Apr 07 15:04:57.215 ERROR vector::unit_test: Failed to execute tests/behavior/transforms/ansi_stripper.toml tests:
invalid escape character in string: `0` at line 25
```

But the `strip_ansi_escapes` creates give this example using bytes:

```rust
fn work() -> io::Result<()> {
  let bytes_with_colors = b"\x1b[32mfoo\x1b[m bar";
  let plain_bytes = strip_ansi_escapes::strip(&bytes_with_colors)?;
  io::stdout().write_all(&plain_bytes)?;
  Ok(())
}
```

**Comment by JeanMertz** [comment]
Thanks for the reproducible example @KannarFr. I'll take a look at this sometime tomorrow.

**Comment by KannarFr** [comment]
@JeanMertz (just a ping :p) maybe i can help you?

**Comment by JeanMertz** [comment]
@KannarFr apologies. I haven't found the time yet, but it's still on my list to do.

What I plan on doing is to check if there's something in the test set-up for this component that differs from your test-case:

https://github.com/timberio/vector/blob/5f78ebb7c3e21f878859b72dff6450c45c7359dc/src/transforms/ansi_stripper.rs#L109-L129

If so, I want to capture your example in a test, and see what's going on.

**Comment by KannarFr** [comment]
I think here if the difference, the test you quoted here doesn't work if you pass its values to a vector test using this transform. As you said.

I think it's a parsing issue, bytes vs String or something like that. 

Waiting for your feedback.

**Comment by JeanMertz** [comment]
Hey @KannarFr,

I discussed this with @LucioFranco. The problem is that the input string uses the octal escape character `\033`, which [isn't supported](https://doc.rust-lang.org/reference/tokens.html#ascii-escapes) by Rust natively. For this to be a valid escape character, we'll need to rewrite it.

Basically something similar to what happens in the `rust-cexpr` crate, [here](https://github.com/jethrogb/rust-cexpr/blob/2979eee12f92d1903a53f820664c65735b010e0e/src/literal.rs#L182-L202).

Luckily, we're already using the Nom crate in Vector, so we can mostly duplicate that piece of code just for this specific use-case of handling the octal escape character.

I won't get to it this week, but it's scheduled for next.

**Comment by KannarFr** [comment]
Ok cool, thanks :).

**Comment by KannarFr** [comment]
Can I help you @JeanMertz?

**Comment by KannarFr** [comment]
:)

**Comment by JeanMertz** [comment]
Hey @KannarFr, I wasn't able to get into this due to other work I'm focused on, but @ktff should be able to help you. I'm not sure when they have time to look into this, but I'll ask them for an update.

**Comment by ktff** [comment]
Hi @KannarFr , I'll try to look into it today, or at worst tomorrow.

**Comment by ktff** [comment]
>  The problem is that the input string uses the octal escape character \033

This is true, but so is this



> @JeanMertz I think there is an issue on the parsing:
> 
> ```
> value = "\x1b[32mfoo\x1b[m bar"
> ```
> 
> throws:
> 
> ```
> Apr 07 15:07:22.938 ERROR vector::unit_test: Failed to execute tests/behavior/transforms/ansi_stripper.toml tests:
> invalid escape character in string: `x` at line 25
> ```

While the problem with octal is fixable as mentioned in [this comment](https://github.com/timberio/vector/issues/2202#issuecomment-612110927), the second one is tougher.

The second problem occurs while decoding `Config` from `toml`. It's the `deserializer` who doesn't recognize the escape characters. To solve it we could open/push a feature request to add it to `serde`/`toml`.

While the problem with octal is a code/behavior problem in `ansi_stripper` and is fixable there, the second problem is more general and apllies to all our `toml` configurations and probably can't be fixed from our side, as such I'll open a separate issue for this second problem, and a PR for the first on.

**Comment by ktff** [comment]
Regarding the octal escape characters, it's actually working as intended. `ansi_stripper` is removing not printable and ANSI Escape sequences, so `ansi_stripper` actually doesn't deal with any `\` based escapes but with the actual characters like newline, tab, escape, etc. So it doesn't remove `\t`, `\r`, `\x1b`, etc.  

The confusion arouse from different approaches that `ansi_stripper`, Rust, and `serde`, have towards escape characters. Rust will in all but raw mode convert `\t`, `\r`, `\x1b`, etc. to the actual characters, so `ansi_stripper` will remove them,  but won't `\033` since Rust wont convert it. And the issue with `serde` is covered with #2492 .

---

In short, there is no bug here, just a missing feature #2492 which only affects `toml` configuration. As such, I'll close this issue.

**Comment by binarylogic** [comment]
To clarify, this works in practice, it just cannot be tested directly with Vector's configuration based unit tests, correct?

**Comment by KannarFr** [comment]
Well, will retry by side but I wrote this test to ensure the issue I was having using vector directly was correct.

**Comment by binarylogic** [comment]
I agree, our tests should support this, but according to the above comments the transform should work as expected.

**Comment by KannarFr** [comment]
I mean I'm pretty sure it affects the runtime too. And to illustrate it I created the issue. Anyway will retry asap and come back to you.

**Comment by ktff** [comment]
> To clarify, this works in practice, it just cannot be tested directly with Vector's configuration based unit tests, correct?

@binarylogic Yes. 

And to further clarify, testing with octal escape sequence `\033` through the Rust code with for example 

```
  let bytes_with_colors = b"\033[32mfoo bar";
  let plain_bytes = strip_ansi_escapes::strip(&bytes_with_colors)?;  
  assert_eq!(plain_bytes.as_slice(),"foo bar".as_bytes());
```

will fail not because `trip_ansi_escapes::strip` is not working, but because `rustc` will "\033" part of the string interpret as `null` character followed by characters "33". So the `bytes_with_colors` will actually contain byte `0` followed with bytes for "33[32mfoo bar". So `plain_bytes` will actually be "33[32mfoo bar", which is correct since there is no ANSI escape sequences in there and `null`, as a non printable character, has been removed.
